testsuite/smoke-test/src/key_manager.rs

```diff
@@ -1,69 +1,74 @@
-        key_manager_config.secure_backend = storage_backend.clone();
-        (&storage_backend).try_into().unwrap()
-    } else {
-        panic!("Couldn't load identity from storage");
-    };
-
-    // Save the key manager config to disk
-    let key_manager_config_path = node_config_path.with_file_name("key_manager.yaml");
-    key_manager_config.save(&key_manager_config_path).unwrap();
+    // Load validator's on disk storage
+    let secure_backend = load_backend_storage(&&node_config);
+    let storage: Storage = (&secure_backend).try_into().unwrap();
 
     // Create a json-rpc connection to the blockchain and verify storage matches the on-chain state.
-    let libra_interface = JsonRpcLibraInterface::new(json_rpc_endpoint);
+    let json_rpc_endpoint = format!("http://127.0.0.1:{}", node_config.rpc.address.port());
+    let libra_interface = JsonRpcLibraInterface::new(json_rpc_endpoint.clone());
     let account = node_config.validator_network.unwrap().peer_id();
     let current_consensus = storage.get_public_key(CONSENSUS_KEY).unwrap().public_key;
     let validator_info = libra_interface.retrieve_validator_info(account).unwrap();
     assert_eq!(&current_consensus, validator_info.consensus_public_key());
 
-    // Spawn the key manager and sleep until a rotation occurs.
-    let mut command = Command::new(workspace_builder::get_bin(KEY_MANAGER_BIN));
-    command
-        .current_dir(workspace_builder::workspace_root())
-        .arg(key_manager_config_path);
+    // Create the key manager
+    let key_manager_storage: Storage = (&secure_backend).try_into().unwrap();
+    let mut key_manager = KeyManager::new(
+        JsonRpcLibraInterface::new(json_rpc_endpoint),
+        key_manager_storage,
+        key_manager_time,
+        1,
+        1000, // Large sleep period to force a single rotation
+        1000,
+        ChainId::test(),
+    );
 
-    let mut key_manager = command.spawn().unwrap();
-    sleep(Duration::from_secs(20));
+    // Spawn the key manager and execute a rotation
+    sleep(Duration::from_secs(10)); // Wait a few seconds for the key to become stale
+    let key_manager_thread = thread::spawn(move || key_manager.execute());
+    sleep(Duration::from_secs(10)); // Wait a few seconds for the blockchain to update
 
     // Verify the consensus key has been rotated in secure storage and on-chain.
     let rotated_consensus = storage.get_public_key(CONSENSUS_KEY).unwrap().public_key;
     let validator_info = libra_interface.retrieve_validator_info(account).unwrap();
     assert_eq!(&rotated_consensus, validator_info.consensus_public_key());
     assert_ne!(current_consensus, rotated_consensus);
 
-    // Kill the key manager process
-    key_manager.kill().unwrap();
+    // Kill the key manager thread
+    unsafe {
+        let raw_handle = key_manager_thread.as_pthread_t();
+        let result = pthread_cancel(raw_handle);
+        assert_eq!(0, result);
+    }
```

Comment:
I would suggest avoiding this as most things i've read up on "killing" threads leads me to believe its a mine field due to invariants being explicitly broken. It's also not portable IIUC.

If we want to go down this route we probably should introduce a way for cleanly notifying the service running in the thread that it should shut down.